presto-main/src/main/java/com/facebook/presto/server/remotetask/ReactorNettyHttpClient.java

```diff
@@ -0,0 +1,385 @@
+                    throw new IllegalArgumentException("KeyStore file path is unreadable or doesn't exist");
+                }
+                if (!Files.exists(trustCertificateFile.toPath()) || !Files.isReadable(trustCertificateFile.toPath())) {
+                    throw new IllegalArgumentException("TrustStore file path is unreadable or doesn't exist");
+                }
+                PrivateKey privateKey = loadPrivateKey(keyFile, Optional.of(internalCommunicationConfig.getKeyStorePassword()));
+                X509Certificate[] certificateChain = readCertificateChain(keyFile).toArray(new X509Certificate[0]);
+                X509Certificate[] trustChain = readCertificateChain(trustCertificateFile).toArray(new X509Certificate[0]);
+
+                String os = System.getProperty("os.name");
+                if (os.toLowerCase(Locale.ENGLISH).contains("linux")) {
+                    // Make sure Open ssl is available for linux deployments
+                    if (!OpenSsl.isAvailable()) {
+                        throw new UnsupportedOperationException(format("OpenSsl is not unavailable. Stacktrace: %s", Arrays.toString(OpenSsl.unavailabilityCause().getStackTrace()).replace(',', '\n')));
+                    }
+                    // Make sure epoll threads are used for linux deployments
+                    if (!Epoll.isAvailable()) {
+                        throw new UnsupportedOperationException(format("Epoll is not unavailable. Stacktrace: %s", Arrays.toString(Epoll.unavailabilityCause().getStackTrace()).replace(',', '\n')));
+                    }
+                }
+
+                SslProvider provider = isAlpnSupported(OPENSSL) ? OPENSSL : JDK;
+                SslContextBuilder sslContextBuilder = SslContextBuilder.forClient()
+                        .sslProvider(provider)
+                        .protocols(TLS_v1_3, TLS_v1_2)
+                        .keyManager(privateKey, certificateChain)
+                        .trustManager(trustChain)
+                        .applicationProtocolConfig(new ApplicationProtocolConfig(ALPN, NO_ADVERTISE, ACCEPT, HTTP_2, HTTP_1_1));
+                if (internalCommunicationConfig.getIncludedCipherSuites().isPresent()) {
+                    sslContextBuilder.ciphers(Splitter
+                            .on(',')
+                            .trimResults()
+                            .omitEmptyStrings()
+                            .splitToList(internalCommunicationConfig.getIncludedCipherSuites().get()));
+                }
+
+                sslContext = sslContextBuilder.build();
+            }
+            catch (IOException | GeneralSecurityException e) {
+                throw new RuntimeException("Failed to configure SSL context", e);
+            }
+        }
+
+        // Create base connection provider with limits
+        ConnectionProvider pool = ConnectionProvider.builder("http-pool").maxConnections(config.getMaxConnections()).build();
```

Comment:
We need to somehow find a way to keep the minimum number of connections setting for HTTP/2. Otherwise all requests will be sent over a single connection handled by a single CPU when HTTP/2 is enabled

Reply:
hmm only one connection is used until the request count reaches the max stream limit right.. after that a new connection will be spawned right ?
may be we can keep the maxIdleTime as high value.. so that a connection will not be closed if its idle for small time.. ?

Reply:
So looks like we can still specify the http2 allocation strategy even if the ConnectionProvider is used for both http1 and http2 connections. 
The ConnectionProvider is like wrapper and underlying there is a separate pool of connections for http1 and http2 protocols. Basically different pools for different protocols. Reactor Netty's HttpConnectionProvider will wrap this connection provider and handle protocol routing in the acquire() call. It examines the configured protocols and routes requests appropriately. So the http2 allocation strategy defined here will only be used for http2 connections.